File: kahoot_server.cpp
// kahoot_server.cpp
// Build with: g++ -std=c++17 kahoot_server.cpp -lboost_system -lpthread
// Make sure Boost headers/libs are installed.

#include <boost/asio.hpp>
#include <boost/beast.hpp>
#include <boost/beast/websocket.hpp>
#include <boost/asio/steady_timer.hpp>
#include <iostream>
#include <string>
#include <memory>
#include <set>
#include <map>
#include <deque>
#include <nlohmann/json.hpp>


using tcp = boost::asio::ip::tcp;
namespace websocket = boost::beast::websocket;
using json = nlohmann::json;
using namespace std::chrono_literals;

class Session : public std::enable_shared_from_this<Session> {
public:
    websocket::stream<tcp::socket> ws;
    std::string name;
    int player_id = -1;
    Session(tcp::socket socket) : ws(std::move(socket)) {}

    void start(std::function<void(std::shared_ptr<Session>, const json&)> on_msg,
               std::function<void(std::shared_ptr<Session>)> on_close) {
        on_message = on_msg;
        on_close_cb = on_close;
        ws.async_accept([self=shared_from_this()](boost::beast::error_code ec){
            if(ec) return;
            self->do_read();
        });
    }

    void send(const json& j) {
        auto txt = j.dump();
        // enqueue to avoid concurrent writes
        write_queue.push_back(txt);
        if(write_queue.size() == 1) do_write();
    }

private:
    std::deque<std::string> write_queue;
    boost::beast::flat_buffer buffer;
    std::function<void(std::shared_ptr<Session>, const json&)> on_message;
    std::function<void(std::shared_ptr<Session>)> on_close_cb;

    void do_read() {
        auto self = shared_from_this();
        ws.async_read(buffer, [self](boost::beast::error_code ec, std::size_t bytes){
            if(ec == boost::beast::websocket::error::closed) {
                if(self->on_close_cb) self->on_close_cb(self);
                return;
            }
            if(ec) {
                std::cerr << "read error: " << ec.message() << "\n";
                if(self->on_close_cb) self->on_close_cb(self);
                return;
            }
            std::string s = boost::beast::buffers_to_string(self->buffer.data());
            self->buffer.consume(self->buffer.size());
            try {
                auto j = json::parse(s);
                if(self->on_message) self->on_message(self, j);
            } catch(std::exception& e) {
                std::cerr << "json parse error: " << e.what() << " data=[" << s << "]\n";
            }
            self->do_read();
        });
    }

    void do_write() {
        auto self = shared_from_this();
        ws.async_write(boost::asio::buffer(write_queue.front()), [self](boost::beast::error_code ec, std::size_t){
            if(ec) {
                std::cerr << "write error: " << ec.message() << "\n";
                return;
            }
            self->write_queue.pop_front();
            if(!self->write_queue.empty()) self->do_write();
        });
    }
};

// Simple in-memory room
struct Player { std::shared_ptr<Session> sess; std::string name; int id; int points = 0; };

class GameRoom {
public:
    std::string code;
    std::map<int, Player> players;
    int next_player_id = 1;
    boost::asio::io_context& ioc;
    GameRoom(boost::asio::io_context& ioc_, const std::string& code_): ioc(ioc_), code(code_) {}

    void broadcast(const json& j) {
        for(auto & [id, p] : players) p.sess->send(j);
    }

    int add_player(std::shared_ptr<Session> s, const std::string& name) {
        int id = next_player_id++;
        players[id] = Player{ s, name, id };
        players[id].sess->player_id = id;
        players[id].sess->name = name;
        json j = { {"type","player_joined"}, {"id",id}, {"name",name} };
        broadcast(j);
        return id;
    }

    void remove_player(std::shared_ptr<Session> s) {
        int id = s->player_id;
        if(id > 0 && players.count(id)) {
            players.erase(id);
            json j = { {"type","player_left"}, {"id",id} };
            broadcast(j);
        }
    }

    // start a question: broadcast question JSON and start timer
    void start_question(int question_id, const json& question_payload, int duration_ms) {
        // add question id and duration
        json msg = question_payload;
        msg["type"] = "question";
        msg["question_id"] = question_id;
        msg["duration_ms"] = duration_ms;
        current_question_id = question_id;
        answers_received.clear();
        broadcast(msg);

        // set a timer to end the question
        timer = std::make_unique<boost::asio::steady_timer>(ioc, std::chrono::milliseconds(duration_ms));
        timer->async_wait([this](const boost::system::error_code& ec){
            if(ec) return;
            this->end_question();
        });
    }

    // handle player answer
    void handle_answer(int player_id, int question_id, int choice, int time_left_ms) {
        if(question_id != current_question_id) return;
        if(answers_received.count(player_id)) return; // ignore double answers
        answers_received[player_id] = {choice, time_left_ms};
        // immediate ack to player
        players[player_id].sess->send(json{{"type","answer_ack"},{"ok",true}});
        // optional: if all answered, end early
        if((int)answers_received.size() >= (int)players.size()) {
            if(timer) timer->cancel();
            end_question();
        }
    }

    void end_question() {
        // compute stats & scoring (example: assume answer 1 is correct)
        int correct_choice = 1; // for demo; replace with real data
        std::vector<int> counts(4,0);
        for(auto &kv : answers_received) {
            int choice = kv.second.choice;
            if(choice >=0 && choice<(int)counts.size()) counts[choice]++;
        }
        // scoring
        for(auto &kv : answers_received) {
            int pid = kv.first;
            int choice = kv.second.choice;
            int time_left_ms = kv.second.time_left_ms;
            if(choice == correct_choice) {
                int base = 1000;
                int bonus = std::max(0, time_left_ms * 1000 / 100); // simplified
                players[pid].points += base + bonus;
            }
        }
        // broadcast result & leaderboard
        json stats = { {"type","question_ended"}, {"counts", counts} };
        broadcast(stats);
        json lb;
        lb["type"] = "leaderboard";
        lb["top"] = json::array();
        for(auto &pr : players) {
            lb["top"].push_back({ {"id", pr.first}, {"name", pr.second.name}, {"points", pr.second.points} });
        }
        broadcast(lb);
    }

private:
    struct Answer { int choice; int time_left_ms; };
    std::map<int, Answer> answers_received;
    int current_question_id = -1;
    std::unique_ptr<boost::asio::steady_timer> timer;
};

int main() {
    try {
        boost::asio::io_context ioc{1};
        tcp::acceptor acceptor{ioc, tcp::endpoint{tcp::v4(), 8080}};

        // single demo room
        auto room = std::make_shared<GameRoom>(ioc, "ABCD");

        std::function<void()> do_accept;
        do_accept = [&](){
            acceptor.async_accept([&](boost::system::error_code ec, tcp::socket socket){
                if(ec) {
                    std::cerr << "accept error: " << ec.message() << "\n";
                    do_accept();
                    return;
                }
                auto session = std::make_shared<Session>(std::move(socket));
                session->start(
                    // on message
                    [&](std::shared_ptr<Session> s, const json& j){
                        try {
                            std::string t = j.value("type", "");
                            if(t == "join") {
                                std::string name = j.value("name","guest");
                                int id = room->add_player(s, name);
                                s->send(json{{"type","joined"},{"id",id},{"room",room->code}});
                            } else if(t == "answer") {
                                int qid = j.value("question_id",-1);
                                int choice = j.value("choice", -1);
                                int time_left = j.value("time_left_ms", 0);
                                room->handle_answer(s->player_id, qid, choice, time_left);
                            } else if(t == "start_question") {
                                int qid = j.value("question_id", 1);
                                json q = {
                                    {"text","What's 2+2?"},
                                    {"choices", json::array({"1","3","4","5"})}
                                };
                                room->start_question(qid, q, 15000);
                            }
                        } catch(...) {}
                    },
                    // on close
                    [&](std::shared_ptr<Session> s){
                        room->remove_player(s);
                    }
                );
                do_accept();
            });
        };
        do_accept();

        std::cout << "Server running on :8080\n";
        ioc.run();
    } catch(std::exception& e) {
        std::cerr << "Fatal: " << e.what() << "\n";
    }
    return 0;
}

File: server/kahoot_server.py
# kahoot_server.py
# Run: python3 kahoot_server.py
# Requires: pip install websockets

import asyncio
import json
import traceback
from collections import deque

class Session:
    """
    One websocket connection. Holds player_id and name after joining.
    """
    def __init__(self, websocket):
        self.ws = websocket
        self.name = ""
        self.player_id = -1
        self._write_queue = deque()
        self._write_lock = asyncio.Lock()

    async def send(self, j):
        txt = json.dumps(j)
        async with self._write_lock:
            await self.ws.send(txt)

class Player:
    def __init__(self, sess, name, pid):
        self.sess = sess
        self.name = name
        self.id = pid
        self.points = 0

class GameRoom:
    def __init__(self, code):
        self.code = code
        self.players = {}
        self.next_player_id = 1
        self.current_question_id = -1
        self.answers_received = {}
        self.timer_task = None
        self.gameStarted = False

    async def broadcast(self, j):
        for p in self.players.values():
            try:
                await p.sess.send(j)
            except:
                # ignore failed sends
                pass

    async def add_player(self, s: Session, name: str):
        pid = self.next_player_id
        self.next_player_id += 1
        self.players[pid] = Player(s, name, pid)
        s.player_id = pid
        s.name = name
        await self.broadcast({"type":"player_joined","id":pid,"name":name})
        return pid

    async def remove_player(self, s: Session):
        pid = s.player_id
        if pid > 0 and pid in self.players:
            del self.players[pid]
            self.next_player_id -= 1
            await self.broadcast({"type":"player_left","id":pid})
            if len(self.players) == 0:
                room.end_game(s)

    async def end_game(self, s: Session):
        self.gameStarted = False

    async def start_game(self, s: Session):
        self.current_question_id = 0
        self.gameStarted = True
        #Look in data base and get question 1, for now ill supply dummy question info
        q = {"text":"What's 2+2?","choices":["1","3","4","5"]}
        await room.start_question(self.current_question_id, q, 15000)

    
    async def start_question(self, question_id, question_payload, duration_ms):
        msg = dict(question_payload)
        msg["type"] = "question"
        msg["question_id"] = question_id
        msg["duration_ms"] = duration_ms
        self.current_question_id = question_id
        self.answers_received.clear()
        await self.broadcast(msg)

        # cancel any existing timer
        if self.timer_task and not self.timer_task.done():
            self.timer_task.cancel()

        self.timer_task = asyncio.create_task(self._end_question_after(duration_ms/1000))

    async def _end_question_after(self, seconds):
        try:
            await asyncio.sleep(seconds)
            await self.end_question()
        except asyncio.CancelledError:
            pass

    async def handle_answer(self, player_id, question_id, choice, time_left_ms):
        if question_id != self.current_question_id:
            return
        if player_id in self.answers_received:
            return
        self.answers_received[player_id] = (choice, time_left_ms)
        await self.players[player_id].sess.send({"type":"answer_ack","ok":True})
        if len(self.answers_received) >= len(self.players):
            if self.timer_task:
                self.timer_task.cancel()
            await self.end_question()

    async def end_question(self):
        correct_choice = 1  # for demo
        counts = [0,0,0,0]
        for choice, _ in self.answers_received.values():
            if 0 <= choice < len(counts):
                counts[choice] += 1
        for pid,(choice,time_left_ms) in self.answers_received.items():
            if choice == correct_choice:
                base = 1000
                bonus = max(0, time_left_ms*1000//100)
                self.players[pid].points += base + bonus
        await self.broadcast({"type":"question_ended","counts":counts})
        lb = {"type":"leaderboard","top":[]}
        for pr in self.players.values():
            lb["top"].append({"id":pr.id,"name":pr.name,"points":pr.points})
        await self.broadcast(lb)
        ##question ended and leaderboards sent
        self.current_question_id += 1
        
        asyncio.create_task(self._start_question_after(10)) #start question after 10s
        
    async def _start_question_after(self, seconds):
        try:
            await asyncio.sleep(seconds)
            #Look in data base and get question 1, for now ill supply dummy question info
            q = {"text":"What's 2+2?","choices":["1","3","4","5"]}
            await room.start_question(self.current_question_id, q, 15000)
        except asyncio.CancelledError:
            pass

room = GameRoom("ROOM!")

async def handler(websocket):
    s = Session(websocket)
    try:
        async for raw in websocket:
            try:
                j = json.loads(raw)
            except:
                print("json parse error", raw)
                continue
            t = j.get("type","")
            if t == "join":                                                 #JOIN HANDLER
                name = j.get("name","guest")
                pid = await room.add_player(s, name)
                await s.send({"type":"joined","id":pid,"room":room.code})
            elif t == "answer":                                             ####ANSWER HANDLER
                qid = j.get("question_id",-1)
                choice = j.get("choice",-1)
                time_left = j.get("time_left_ms",0)
                await room.handle_answer(s.player_id,qid,choice,time_left)
            elif t == "player_leave":                                       ####LEAVE HANDLER
                await room.remove_player(s)
            elif t == "start_question":                                     #START QUESTION HANDLER
                qid = j.get("question_id",1)
                q = {"text":"What's 2+2?","choices":["1","3","4","5"]}
                await room.start_question(qid,q,15000)
            elif t == "start_game":                                     #START QUESTION HANDLER
                qid = j.get("question_id",1)
                q = {"text":"What's 2+2?","choices":["1","3","4","5"]}
                await room.start_question(qid,q,15000)
    except:
        traceback.print_exc()


import websockets

async def main():
    print("Server running on :8080")
    async with websockets.serve(handler, "0.0.0.0", 8080):
        await asyncio.Future()  # run forever

if __name__ == "__main__":
    asyncio.run(main())
